#Financial Product Complaint Classification and Summarization

## **Business Context**

### **Description**
*In the modern financial industry, customer complaints play a crucial role in identifying areas where financial institutions can improve their services. Effectively categorizing these complaints into specific product categories, such as credit reports, student loans, or money transfers, is essential for addressing customer concerns promptly by routing the tickets to relevant personnel. Leveraging Generative AI for text classification can help financial institutions better understand customer grievances and respond more efficiently. Apart from this, a summary of the customer complaint helps the support personnel quickly grasp the gist of the grievance*

### **Objective**
*The primary goal of this project is to utilize Generative AI techniques to improve the classification and summarization of customer complaints in the financial sector.
Specifically, the project will focus on:*

1. **Text-to-Label Classification:** *Implementing Zero-shot and Few-shot prompting methods to accurately classify customer complaints into relevant product categories.*
2. **Text-to-Text Summarization:** *Using Zero-shot prompting to generate concise summaries of customer complaints, enabling more personalized and effective responses.*


# **Section 1 : Setting Up for Prompt Engineering with Mistral Model**

### **Install & Importing neccessary libraries**

In [5]:
!apt-get update
!apt-get install -y ninja-build cmake
!pip install ipywidgets --upgrade


0% [Working]
            
Hit:1 https://cli.github.com/packages stable InRelease

0% [Connecting to archive.ubuntu.com] [Connecting to security.ubuntu.com (91.18
                                                                               
Hit:2 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease
Hit:3 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease
Hit:4 http://security.ubuntu.com/ubuntu jammy-security InRelease
Hit:5 http://archive.ubuntu.com/ubuntu jammy InRelease
Hit:6 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease
Hit:7 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease
Hit:8 http://archive.ubuntu.com/ubuntu jammy-updates InRelease
Hit:9 https://r2u.stat.illinois.edu/ubuntu jammy InRelease
Hit:10 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease
Hit:11 http://archive.ubuntu.com/ubuntu jammy-backports InRelease
Reading package lists... Done
W: Skipping acq

In [6]:
import torch

In [7]:
# This part of code will skip all the un-necessary warnings which can occur during the execution of this project.
import warnings
warnings.filterwarnings("ignore", category=Warning)

In [8]:
# Installation for GPU llama-cpp-python==0.2.69
!CMAKE_ARGS="-DLLAMA_CUDA=on" pip install llama-cpp-python==0.2.69
# For downloading the models from HF Hub
!pip install huggingface_hub

In [9]:
!pip install evaluate #HF to evaluate metrics for ML models
!pip install bert-score #how similar 2 sentences are using contextual embeddings from BERT like models

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.1/84.1 kB 8.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.1/61.1 kB 6.2 MB/s eta 0:00:00


In [10]:
!pip freeze > requirement.txt

In [11]:
# Basic Imports for Libraries
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score

import pandas as pd
import numpy as np
from tqdm import tqdm
import json
import re

import torch
import evaluate

# from google.colab import drive
import locale

###1: Importing Libaries and Mistral Model

In [12]:
from huggingface_hub import hf_hub_download
from llama_cpp import Llama

In [13]:
model_name_or_path = "TheBloke/Mistral-7B-Instruct-v0.2-GGUF"
model_basename = "mistral-7b-instruct-v0.2.Q5_K_M.gguf"

In [14]:
model_path = hf_hub_download(
    repo_id=model_name_or_path,
    filename=model_basename
    )

mistral-7b-instruct-v0.2.Q5_K_M.gguf:   0%|          | 0.00/5.13G [00:00<?, ?B/s]

In [15]:
lcpp_llm = Llama(
        model_path=model_path, #GGUF model file # mistral-7b
        n_threads=2,  # no CPU cores used
        n_batch=512,  # tokens processed per computation step
        n_gpu_layers=43,  # how many model layers are offloaded to GPU
        n_ctx=4096,  # context window size
    )

llama_model_loader: loaded meta data with 24 key-value pairs and 291 tensors from /root/.cache/huggingface/hub/models--TheBloke--Mistral-7B-Instruct-v0.2-GGUF/snapshots/3a6fbf4a41a1d52e415a4958cde6856d34b2db93/mistral-7b-instruct-v0.2.Q5_K_M.gguf (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = llama
llama_model_loader: - kv   1:                               general.name str              = mistralai_mistral-7b-instruct-v0.2
llama_model_loader: - kv   2:                       llama.context_length u32              = 32768
llama_model_loader: - kv   3:                     llama.embedding_length u32              = 4096
llama_model_loader: - kv   4:                          llama.block_count u32              = 32
llama_model_loader: - kv   5:                  llama.feed_forward_length u32              = 14336
llama_model_loa

# **Text to Label generation**

In [46]:
system_message = """
You are an expert financial complaint classification assistant.
Your task is to read a customer's complaint narrative and classify it into the correct product category.

Choose only one category from the following list (return exactly as written):
credit_card, retail_banking, credit_reporting, mortgages_and_loans, debt_collection.

Return only the category name. Do not provide explanations.
"""

In [47]:
zero_shot_prompt_template = """<s>[INST]
<<SYS>>
{system_message}
<</SYS>>

Customer Complaint:
{user_input}

Product Category:
[/INST]

"""

In [18]:
def generate_prompt(system_message,user_input):
    prompt=zero_shot_prompt_template.format(system_message=system_message,user_input=user_input) #zero shot prompt
    return prompt

In [19]:
def generate_mistral_response(input_text):

    # Combine user_prompt and system_message to create the prompt
    prompt = generate_prompt(system_message,input_text)

    # Define the Llama model along with its parameters for generating a response
    response = lcpp_llm(
        prompt=prompt,
        max_tokens=1200,
        temperature=0,
        top_p = 0.95,
        repeat_penalty=1.2,
        top_k=50,
        stop=["</s>"],
        echo=False
    )

    # Extract and return the response text
    response_text = response["choices"][0]["text"].strip()
    print(response_text)
    return response_text

In [21]:
data = pd.read_csv("/content/Complaints_classification.csv")

In [22]:
# Randomly select 30 rows
new_data = data.sample(n=30, random_state=40)

In [23]:
print(new_data.head())

              product                                          narrative  \
167    retail_banking  fraudulent charge totaling made capital one ch...   
169  credit_reporting  block except otherwise provided section consum...   
461       credit_card  usaa master plan collect cancellation debt usa...   
253  credit_reporting  block except otherwise provided section consum...   
42   credit_reporting  open account acct opened balance account acct ...   

                                               summary  
167  A fraudulent charge was made on the individual...  
169  The text outlines various stipulations regardi...  
461  The input appears to be a complaint about USAA...  
253  The text pertains to the stipulations and oper...  
42   The input is about various accounts being open...  


In [24]:
#create a new column DF 'mistral_response' & populate it with responses generated by applying the 'general_mistral_response' function to each 'narrative' in the DF & prepare the mistral_response_cleaned col using extract_category function
new_data['mistral_response'] = new_data['narrative'].apply(lambda x: generate_mistral_response(x))



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       3.97 ms /     6 runs   (    0.66 ms per token,  1510.19 tokens per second)
llama_print_timings: prompt eval time =     649.20 ms /   338 tokens (    1.92 ms per token,   520.64 tokens per second)
llama_print_timings:        eval time =     139.62 ms /     5 runs   (   27.92 ms per token,    35.81 tokens per second)
llama_print_timings:       total time =     817.65 ms /   343 tokens
Llama.generate: prefix-match hit


credit_card.



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       3.17 ms /     6 runs   (    0.53 ms per token,  1893.34 tokens per second)
llama_print_timings: prompt eval time =     513.79 ms /   512 tokens (    1.00 ms per token,   996.52 tokens per second)
llama_print_timings:        eval time =     150.38 ms /     5 runs   (   30.08 ms per token,    33.25 tokens per second)
llama_print_timings:       total time =     684.44 ms /   517 tokens
Llama.generate: prefix-match hit


credit_reporting



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       3.27 ms /     6 runs   (    0.54 ms per token,  1835.42 tokens per second)
llama_print_timings: prompt eval time =     320.88 ms /   189 tokens (    1.70 ms per token,   589.01 tokens per second)
llama_print_timings:        eval time =     146.58 ms /     5 runs   (   29.32 ms per token,    34.11 tokens per second)
llama_print_timings:       total time =     488.83 ms /   194 tokens
Llama.generate: prefix-match hit


debt_collection.



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       3.31 ms /     6 runs   (    0.55 ms per token,  1813.24 tokens per second)
llama_print_timings: prompt eval time =     513.65 ms /   512 tokens (    1.00 ms per token,   996.78 tokens per second)
llama_print_timings:        eval time =     154.43 ms /     5 runs   (   30.89 ms per token,    32.38 tokens per second)
llama_print_timings:       total time =     688.61 ms /   517 tokens
Llama.generate: prefix-match hit


credit_reporting



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       2.72 ms /     6 runs   (    0.45 ms per token,  2202.64 tokens per second)
llama_print_timings: prompt eval time =     200.57 ms /    81 tokens (    2.48 ms per token,   403.86 tokens per second)
llama_print_timings:        eval time =     148.18 ms /     5 runs   (   29.64 ms per token,    33.74 tokens per second)
llama_print_timings:       total time =     368.91 ms /    86 tokens
Llama.generate: prefix-match hit


retail_banking



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       3.30 ms /     6 runs   (    0.55 ms per token,  1820.39 tokens per second)
llama_print_timings: prompt eval time =     487.54 ms /   427 tokens (    1.14 ms per token,   875.83 tokens per second)
llama_print_timings:        eval time =     154.14 ms /     5 runs   (   30.83 ms per token,    32.44 tokens per second)
llama_print_timings:       total time =     662.14 ms /   432 tokens
Llama.generate: prefix-match hit


credit_reporting



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       5.51 ms /    10 runs   (    0.55 ms per token,  1815.54 tokens per second)
llama_print_timings: prompt eval time =     206.95 ms /    94 tokens (    2.20 ms per token,   454.21 tokens per second)
llama_print_timings:        eval time =     268.05 ms /     9 runs   (   29.78 ms per token,    33.58 tokens per second)
llama_print_timings:       total time =     507.39 ms /   103 tokens
Llama.generate: prefix-match hit


mortgages_and_loans



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       3.44 ms /     6 runs   (    0.57 ms per token,  1742.16 tokens per second)
llama_print_timings: prompt eval time =     486.62 ms /   427 tokens (    1.14 ms per token,   877.47 tokens per second)
llama_print_timings:        eval time =     155.28 ms /     5 runs   (   31.06 ms per token,    32.20 tokens per second)
llama_print_timings:       total time =     664.63 ms /   432 tokens
Llama.generate: prefix-match hit


credit_reporting



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       3.20 ms /     6 runs   (    0.53 ms per token,  1877.35 tokens per second)
llama_print_timings: prompt eval time =     487.06 ms /   439 tokens (    1.11 ms per token,   901.33 tokens per second)
llama_print_timings:        eval time =     151.68 ms /     5 runs   (   30.34 ms per token,    32.97 tokens per second)
llama_print_timings:       total time =     658.79 ms /   444 tokens
Llama.generate: prefix-match hit


credit_reporting



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       3.04 ms /     6 runs   (    0.51 ms per token,  1974.33 tokens per second)
llama_print_timings: prompt eval time =     488.45 ms /   427 tokens (    1.14 ms per token,   874.20 tokens per second)
llama_print_timings:        eval time =     156.31 ms /     5 runs   (   31.26 ms per token,    31.99 tokens per second)
llama_print_timings:       total time =     664.33 ms /   432 tokens
Llama.generate: prefix-match hit


credit_reporting



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       3.66 ms /     6 runs   (    0.61 ms per token,  1638.45 tokens per second)
llama_print_timings: prompt eval time =     267.95 ms /   154 tokens (    1.74 ms per token,   574.74 tokens per second)
llama_print_timings:        eval time =     149.85 ms /     5 runs   (   29.97 ms per token,    33.37 tokens per second)
llama_print_timings:       total time =     448.19 ms /   159 tokens
Llama.generate: prefix-match hit


credit_reporting



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       3.39 ms /     6 runs   (    0.56 ms per token,  1771.48 tokens per second)
llama_print_timings: prompt eval time =     601.64 ms /   518 tokens (    1.16 ms per token,   860.98 tokens per second)
llama_print_timings:        eval time =     159.39 ms /     5 runs   (   31.88 ms per token,    31.37 tokens per second)
llama_print_timings:       total time =     785.54 ms /   523 tokens
Llama.generate: prefix-match hit


credit_reporting



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       3.03 ms /     6 runs   (    0.51 ms per token,  1977.59 tokens per second)
llama_print_timings: prompt eval time =     489.16 ms /   427 tokens (    1.15 ms per token,   872.93 tokens per second)
llama_print_timings:        eval time =     154.78 ms /     5 runs   (   30.96 ms per token,    32.30 tokens per second)
llama_print_timings:       total time =     663.86 ms /   432 tokens
Llama.generate: prefix-match hit


credit_reporting



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       5.69 ms /    10 runs   (    0.57 ms per token,  1756.23 tokens per second)
llama_print_timings: prompt eval time =     296.83 ms /   239 tokens (    1.24 ms per token,   805.19 tokens per second)
llama_print_timings:        eval time =     277.27 ms /     9 runs   (   30.81 ms per token,    32.46 tokens per second)
llama_print_timings:       total time =     610.38 ms /   248 tokens
Llama.generate: prefix-match hit


mortgages_and_loans



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       3.11 ms /     6 runs   (    0.52 ms per token,  1930.50 tokens per second)
llama_print_timings: prompt eval time =     587.11 ms /   518 tokens (    1.13 ms per token,   882.29 tokens per second)
llama_print_timings:        eval time =     157.41 ms /     5 runs   (   31.48 ms per token,    31.76 tokens per second)
llama_print_timings:       total time =     765.93 ms /   523 tokens
Llama.generate: prefix-match hit


credit_reporting



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       4.05 ms /     6 runs   (    0.68 ms per token,  1480.02 tokens per second)
llama_print_timings: prompt eval time =     493.65 ms /   445 tokens (    1.11 ms per token,   901.45 tokens per second)
llama_print_timings:        eval time =     151.46 ms /     5 runs   (   30.29 ms per token,    33.01 tokens per second)
llama_print_timings:       total time =     678.38 ms /   450 tokens
Llama.generate: prefix-match hit


credit_reporting



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       4.03 ms /     6 runs   (    0.67 ms per token,  1489.57 tokens per second)
llama_print_timings: prompt eval time =     490.54 ms /   427 tokens (    1.15 ms per token,   870.48 tokens per second)
llama_print_timings:        eval time =     152.95 ms /     5 runs   (   30.59 ms per token,    32.69 tokens per second)
llama_print_timings:       total time =     677.60 ms /   432 tokens
Llama.generate: prefix-match hit


credit_reporting



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       4.27 ms /     6 runs   (    0.71 ms per token,  1405.48 tokens per second)
llama_print_timings: prompt eval time =     289.16 ms /   200 tokens (    1.45 ms per token,   691.66 tokens per second)
llama_print_timings:        eval time =     154.02 ms /     5 runs   (   30.80 ms per token,    32.46 tokens per second)
llama_print_timings:       total time =     477.46 ms /   205 tokens
Llama.generate: prefix-match hit


credit_card.



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       4.84 ms /     6 runs   (    0.81 ms per token,  1239.93 tokens per second)
llama_print_timings: prompt eval time =     263.24 ms /   146 tokens (    1.80 ms per token,   554.62 tokens per second)
llama_print_timings:        eval time =     152.96 ms /     5 runs   (   30.59 ms per token,    32.69 tokens per second)
llama_print_timings:       total time =     451.06 ms /   151 tokens
Llama.generate: prefix-match hit


credit_reporting



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       5.80 ms /    10 runs   (    0.58 ms per token,  1723.54 tokens per second)
llama_print_timings: prompt eval time =     404.75 ms /   365 tokens (    1.11 ms per token,   901.79 tokens per second)
llama_print_timings:        eval time =     280.09 ms /     9 runs   (   31.12 ms per token,    32.13 tokens per second)
llama_print_timings:       total time =     720.92 ms /   374 tokens
Llama.generate: prefix-match hit


mortgages_and_loans



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       3.10 ms /     6 runs   (    0.52 ms per token,  1934.86 tokens per second)
llama_print_timings: prompt eval time =     200.86 ms /    79 tokens (    2.54 ms per token,   393.30 tokens per second)
llama_print_timings:        eval time =     151.22 ms /     5 runs   (   30.24 ms per token,    33.07 tokens per second)
llama_print_timings:       total time =     372.90 ms /    84 tokens
Llama.generate: prefix-match hit


credit_reporting



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       3.49 ms /     6 runs   (    0.58 ms per token,  1717.23 tokens per second)
llama_print_timings: prompt eval time =     397.98 ms /   343 tokens (    1.16 ms per token,   861.86 tokens per second)
llama_print_timings:        eval time =     151.63 ms /     5 runs   (   30.33 ms per token,    32.97 tokens per second)
llama_print_timings:       total time =     571.71 ms /   348 tokens
Llama.generate: prefix-match hit


credit_reporting



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       3.25 ms /     6 runs   (    0.54 ms per token,  1845.59 tokens per second)
llama_print_timings: prompt eval time =     215.32 ms /   118 tokens (    1.82 ms per token,   548.02 tokens per second)
llama_print_timings:        eval time =     151.40 ms /     5 runs   (   30.28 ms per token,    33.02 tokens per second)
llama_print_timings:       total time =     386.37 ms /   123 tokens
Llama.generate: prefix-match hit


credit_reporting



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       3.27 ms /     6 runs   (    0.54 ms per token,  1837.67 tokens per second)
llama_print_timings: prompt eval time =     262.05 ms /   140 tokens (    1.87 ms per token,   534.24 tokens per second)
llama_print_timings:        eval time =     156.60 ms /     5 runs   (   31.32 ms per token,    31.93 tokens per second)
llama_print_timings:       total time =     438.54 ms /   145 tokens
Llama.generate: prefix-match hit


debt_collection.



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       3.29 ms /     6 runs   (    0.55 ms per token,  1825.37 tokens per second)
llama_print_timings: prompt eval time =     385.05 ms /   297 tokens (    1.30 ms per token,   771.33 tokens per second)
llama_print_timings:        eval time =     155.05 ms /     5 runs   (   31.01 ms per token,    32.25 tokens per second)
llama_print_timings:       total time =     560.69 ms /   302 tokens
Llama.generate: prefix-match hit


retail_banking



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       3.14 ms /     6 runs   (    0.52 ms per token,  1912.66 tokens per second)
llama_print_timings: prompt eval time =     495.06 ms /   427 tokens (    1.16 ms per token,   862.52 tokens per second)
llama_print_timings:        eval time =     159.14 ms /     5 runs   (   31.83 ms per token,    31.42 tokens per second)
llama_print_timings:       total time =     674.43 ms /   432 tokens
Llama.generate: prefix-match hit


credit_reporting



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       3.48 ms /     6 runs   (    0.58 ms per token,  1722.65 tokens per second)
llama_print_timings: prompt eval time =     213.86 ms /   110 tokens (    1.94 ms per token,   514.36 tokens per second)
llama_print_timings:        eval time =     152.65 ms /     5 runs   (   30.53 ms per token,    32.76 tokens per second)
llama_print_timings:       total time =     422.85 ms /   115 tokens
Llama.generate: prefix-match hit


credit_reporting



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       5.94 ms /     6 runs   (    0.99 ms per token,  1010.78 tokens per second)
llama_print_timings: prompt eval time =     286.81 ms /   159 tokens (    1.80 ms per token,   554.36 tokens per second)
llama_print_timings:        eval time =     143.76 ms /     5 runs   (   28.75 ms per token,    34.78 tokens per second)
llama_print_timings:       total time =     484.38 ms /   164 tokens
Llama.generate: prefix-match hit


credit_reporting



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       3.23 ms /     6 runs   (    0.54 ms per token,  1855.86 tokens per second)
llama_print_timings: prompt eval time =     523.28 ms /   512 tokens (    1.02 ms per token,   978.45 tokens per second)
llama_print_timings:        eval time =     155.76 ms /     5 runs   (   31.15 ms per token,    32.10 tokens per second)
llama_print_timings:       total time =     699.21 ms /   517 tokens
Llama.generate: prefix-match hit


credit_reporting



llama_print_timings:        load time =     649.71 ms
llama_print_timings:      sample time =       3.26 ms /     6 runs   (    0.54 ms per token,  1840.49 tokens per second)
llama_print_timings: prompt eval time =     212.12 ms /   102 tokens (    2.08 ms per token,   480.86 tokens per second)
llama_print_timings:        eval time =     158.42 ms /     5 runs   (   31.68 ms per token,    31.56 tokens per second)
llama_print_timings:       total time =     389.76 ms /   107 tokens


credit_reporting


In [25]:
new_data['mistral_response']

,mistral_response
167,credit_card.
169,credit_reporting
461,debt_collection.
253,credit_reporting
42,retail_banking
369,credit_reporting
26,mortgages_and_loans
377,credit_reporting
238,credit_reporting
374,credit_reporting


In [26]:
#Extracting a category label from text using regular expressions.
#Find a structured label first. If that fails, it falls back to searching for known category keywords.

#Creates a function that takes 1 piece of text as input
def extract_category(text):
    # Define the regex pattern to match "category:" or "Category:" followed by a word
    pattern = r'category:\s*(\w+)'

    # Use re.search with the re.IGNORECASE flag to make it case-insensitive
    match = re.search(pattern, text, re.IGNORECASE)

    # If a match is found, return the captured group, else return None
    if match:
        return match.group(1)
    else:
        pattern1 = r'(credit_card|retail_banking|credit_reporting|mortgages_and_loans|debt_collection)'
        match = re.search(pattern1, text, re.IGNORECASE)
        if match:
            return match.group()
        else:
            return ''

In [27]:
#Handles missing responses
#Extract structured labels like "Category: Credit_Card"
#Fall back to keyword search
#Defaults to "other" when nothing matches

import re

def extract_category(text):
    if not text:
        return "Other"

    # Match "Category: Something With Spaces"
    pattern = r'category:\s*([A-Za-z\s]+)'
    match = re.search(pattern, text, re.IGNORECASE)

    if match:
        return match.group(1).strip()

    # Fallback to known labels
    allowed = [
        "credit_card",
        "retail_banking",
        "credit_reporting",
        "mortgages_and_loans",
        "debt_collection",
        "Other"
    ]

    for cat in allowed:
        if cat.lower() in text.lower():
            return cat

    return "Other"

In [28]:
new_data['mistral_response_cleaned'] = new_data['mistral_response'].apply(lambda x: extract_category(x))

In [29]:
new_data.head()

,product,narrative,summary,mistral_response,mistral_response_cleaned
167,retail_banking,fraudulent charge totaling made capital one ch...,A fraudulent charge was made on the individual...,credit_card.,credit_card
169,credit_reporting,block except otherwise provided section consum...,The text outlines various stipulations regardi...,credit_reporting,credit_reporting
461,credit_card,usaa master plan collect cancellation debt usa...,The input appears to be a complaint about USAA...,debt_collection.,debt_collection
253,credit_reporting,block except otherwise provided section consum...,The text pertains to the stipulations and oper...,credit_reporting,credit_reporting
42,credit_reporting,open account acct opened balance account acct ...,The input is about various accounts being open...,retail_banking,retail_banking


In [30]:
f1 =  f1_score(new_data['product'], new_data['mistral_response_cleaned'],average='micro')
print(f'F1 Score: {f1}')

F1 Score: 0.8333333333333334


Since micro-averaging aggregates all true positives, false positives ,and false negatives globally, this score is equivalent to overall accuracy in a multi-class setting.

In [ ]:
f2 = f1_score(
    new_data['product'],
    new_data['mistral_response_cleaned'],
    average='micro'
)

print(f'F1 Score: {f2}')

F1 Score: 0.8333333333333334


F1 score for both mistral_response and mistral_response_cleaned is 0.8333, indicating that the cleaning step did not change the model's classification and performance.

The cleaning function mainly standardizes the output format of the model predictions (eg, extracting the category label and normalizing text). However, since the predicted categories remained the same before and after cleaning, the number of true positives, false positives, and false negative did not change.

Because the evaluation uses micro-average F1 score, which measures overall classification accuracy across all classes, the resulting score remains identical.

Therefore, the cleaning step improves the consistency of prediction labels but does not affect the model's predicitive performance in this dataset.